### PR TITLE
Fix for huge ballot numbers DDoS

### PR DESCRIPTION
**Pull request by ilya**

### CODE REVIEW

**Comment by ilya** [comment]
reminding myself to double check the gotonextballot logic

File: consensus/ballot_state.go [hunk 1]
```diff
@@ -268,7 +269,7 @@ func (s *BallotState) MaybeConfirmAsPrepared(n int, x SlotValue) bool {
 
 	s.Logf("confirms as prepared: %s", &Ballot{n: n, x: x})
 	
-	if s.hn == n && *s.z != x {
```
Comment (lacker) [inline consensus/ballot_state.go]: why did you remove this check, is it just redundant?

File: consensus/ballot_state.go [hunk 2]
```diff
@@ -529,18 +530,38 @@ func (s *BallotState) RelevantRange(x SlotValue) (int, int) {
...
+	numberToNodes := make(map[int][]string)
+
+	for node, message := range s.M {
+		maxN := message.MaxN()
+		numberToNodes[maxN] = append(numberToNodes[maxN], node)
+	}
+
+	max := 0
+
+	for n, nodes := range numberToNodes {
```
Comment (lacker) [inline consensus/ballot_state.go]: This makes it seem like if you have 3 other nodes and their maxes are 2, 3, and 4, then max will just be 0. That doesn't seem right? It seems like you need to be accumulating or something so that the nodes with a max=4 also count as making n=3 actionable.

File: consensus/block_test.go [hunk 3]
```diff
@@ -81,6 +82,52 @@ func TestConsensus(t *testing.T) {
...
+					}
+				}
+
+				block2.Handle(block.publicKey, message)
+			}
+		}
+	}
+}
+
+func TestProtectionAgainstBigRangeDDoS(t *testing.T) {
```
Comment (lacker) [inline consensus/block_test.go]: does this test infinite-ish-loop the code before your changes?

**Comment by lacker** [comment]
Added some comments. Also, did you test this with something like the chain_test full cluster test with a high loop number?

File: consensus/ballot_state.go (hunk 1 above)
Comment (ilya) [reply consensus/ballot_state.go]: Yeah

File: consensus/block_test.go (hunk 3 above)
Comment (ilya) [reply consensus/block_test.go]: Yep, hangs without the changes.

**Comment by ilya** [comment]
I did with 10k I believe. I’ll run it again.

File: consensus/ballot_state.go (hunk 2 above)
Comment (ilya) [reply consensus/ballot_state.go]: Yeah, that doesn’t seem right. I’ll take a look in a bit.

Comment (lacker) [reply consensus/ballot_state.go]: This might not cause trouble with a full cluster but then cause trouble when there's a knockout because then it's harder to get a blocking set to agree on things. I would try running the knockout tests if you haven't already to see if you can find a deadlock.

Comment (lacker) [reply consensus/ballot_state.go]: (with a higher-than-in-the-code # of loops i mean)

**Comment by ilya** [comment]
Oops, misread your prior comment on my phone. I had only run the block tests with a high loop. Anyhow, this is clearly broken. Working on a fix.

**Comment by ilya** [comment]
Okay, PTAL. Apologies for submitting a broken PR. Sloppy testing on my part.

I've run this one through:
- TestBlockFullCluster w/ 100k
- TestBlockOneNodeKnockedOut w/ 100k
- TestChainFullCluster w/ 10k
- TestChainOneNodeKnockedOut w/ 10k
- The whole suite w/ default loops

I'll try to put together a test script that bumps the loops everywhere and removes the default go timeout. Also they might run faster if we don't dump so much output.

File: consensus/ballot_state.go [hunk 4]
```diff
@@ -529,21 +531,59 @@ func (s *BallotState) RelevantRange(x SlotValue) (int, int) {
...
+// Returns the max ballot number that a blocking set of nodes are talking about.
+func (s *BallotState) MaxActionableBallotNumber() int {
+	numberToNodes := make(map[int][]string)
+
+	for node, message := range s.M {
+		maxN := message.MaxN()
+		numberToNodes[maxN] = append(numberToNodes[maxN], node)
+	}
+
+	i := 0
```
Comment (lacker) [inline consensus/ballot_state.go]: this loop confused me at first, can you avoid the use of i by doing like ns := []int{} and then ns = append(ns, n) inside the loop? i think that is more go-styley

File: consensus/ballot_state.go [hunk 5]
```diff
@@ -529,21 +531,59 @@ func (s *BallotState) RelevantRange(x SlotValue) (int, int) {
...
+
+	sort.Sort(sort.Reverse(sort.IntSlice(nKeys)))
+
+	nodesAbove := []string{}
+	max := 0
+
+	for _, n := range nKeys {
+		nodesAbove = append(nodesAbove, numberToNodes[n]...)
+		if (s.D.BlockedBy(nodesAbove)) {
+			max = n
```
Comment (lacker) [inline consensus/ballot_state.go]: you don't really need max, you can just return n here and return 0 if the for loop terminates

**Comment by lacker** [comment]
Cool, just a couple go-style-y comments

File: consensus/ballot_state.go (hunk 4 above)
Comment (ilya) [reply consensus/ballot_state.go]: I think my approach is more efficient because the size is known a priori.

File: consensus/ballot_state.go (hunk 5 above)
Comment (ilya) [reply consensus/ballot_state.go]: Good point. Will fix.

File: consensus/ballot_state.go (hunk 4 above)
Comment (lacker) [reply consensus/ballot_state.go]: pfff, efficiency when creating an array of size 4. fine

**Comment by lacker** [comment]
LGTM
